dbt_semantic_interfaces/validations/primary_entity.py

```diff
@@ -0,0 +1,101 @@
+    """
+
+    @staticmethod
+    def _model_requires_primary_entity(semantic_model: SemanticModel) -> bool:
+        return len(semantic_model.dimensions) > 0
+
+    @staticmethod
+    def _model_has_entity_with_primary_type(semantic_model: SemanticModel) -> bool:
+        return any(entity.type == EntityType.PRIMARY for entity in semantic_model.entities)
+
+    @staticmethod
+    def _entity_with_primary_type_in_model(semantic_model: SemanticModel) -> Entity:
+        for entity in semantic_model.entities:
+            if entity.type == EntityType.PRIMARY:
+                return entity
+        entities_with_primary_type = tuple(
+            entity for entity in semantic_model.entities if entity.type == EntityType.PRIMARY
+        )
+
+        if len(entities_with_primary_type) == 1:
+            return entities_with_primary_type[0]
+
+        raise RuntimeError(f"Did not find exactly one entity with entity type {EntityType.PRIMARY} in {semantic_model}")
+
+    @staticmethod
+    def _check_model(semantic_model: SemanticModel) -> Sequence[ValidationIssue]:
+        context = SemanticModelContext(
+            file_context=FileContext.from_metadata(metadata=semantic_model.metadata),
+            semantic_model=SemanticModelReference(semantic_model_name=semantic_model.name),
+        )
+
+        # Check that the primary entity field and the listed entities don't conflict.
+        if (
+            semantic_model.primary_entity_reference is not None
+            and PrimaryEntityRule._model_has_entity_with_primary_type(semantic_model)
+        ):
+            entity_with_primary_type = PrimaryEntityRule._entity_with_primary_type_in_model(semantic_model)
+            if semantic_model.primary_entity_reference != entity_with_primary_type.reference:
+                return (
+                    ValidationError(
+                        message=(
+                            f"Semantic model {semantic_model.name} has an entity named {entity_with_primary_type.name} "
+                            f"with type primary but it conflicts with the primary_entity field set to "
+                            f"{semantic_model.primary_entity_reference.element_name}"
+                        ),
+                        context=context,
+                    ),
+                )
+
+        # Check that a primary entity has been set if required.
+        if (
+            PrimaryEntityRule._model_requires_primary_entity(semantic_model)
+            and semantic_model.primary_entity_reference is None
+            and not PrimaryEntityRule._model_has_entity_with_primary_type(semantic_model)
+        ):
+            return (
+                ValidationError(
+                    message=(
+                        f"The semantic model {semantic_model.name} contains dimensions, but it does not define a "
+                        f"primary entity."
+                    ),
+                    context=context,
+                ),
+            )
+
+        return []
```

Comment:
Is it not valid at all for a `SemanticModel` with dimensions to not have a primary entity, or is it something we simply want to discourage? If so should this instead be a `ValidationWarning`? Without a primary entity it just makes the dimensions in accessible to metrics based on different `SemanticModel`s right?

Reply:
It seems like an error to me. If the dimensions are not accessible - why have them defined at all? There could also be some poor UX here - a user makes a model, commits it (allowed because it's a warning), and then wonders why the dimensions aren't showing up.

Reply:
I think this is a misunderstanding on my part. I thought that the `Dimension`s would still be accessible by `Measure`s and `Metric`s that depend on the parent `SemanticModel`. Would it be valid for there to be a simple `Metric` based off a `Measure` in a semantic model that has a time `Dimension` that is the `Measure`s `agg_time_dimension` but no primary entity? An example

```
semantic_modesl:
  - name: user_created_event_log
    dimensions:
      - name: ds
        type: time
        expr: created_at
    measures:
      - name: users_created
        agg: COUNT
        expr: 1

metrics:
  - name: total users
     type: simple
     type_params:
         measure: users_created
```

Reply:
That would need a primary entity since the rule is: if there are dimensions, there needs to be a primary entity.

Reply:
When you say rule here I assume you mean how MetricFlow will be operating on things, not the validation rule itself. So MetricFlow would raise an exception at query time on above if attempting to query the metric `total_users`?